# String
K reprezentaci textu je nejen v Pythonu datový typ string. V Pythonu je to velice jednoduché, pokud chceme vytvořit textový řetězec používáme buď dvojité uvozovky "", nebo jednoduché uvozovky ''. Volba druhu uvozovek záleží pouze na nás, jediné co se musí dodžovat je aby zavádějící uvozovky a ukončující uvozovky byly stejné.

In [ ]:
# ukázka stringu
textovy_retezec = "Maminka říkala " " Vem si kulicha"
print(textovy_retezec)

textovy_retezec = 'Maminka říkala " Vem si kulicha" ' 
print(textovy_retezec)

## Spojení stringů
Může nastat situace kdy budeme mít ve více proměnných datové typy string a budeme je chtít spojit do jednoho stringu. To se děje pomocí operátoru '+'.

In [ ]:
retezec_1 = "Toto"
retezec_2 = "je"
retezec_3 = "spojený"
retezec_4 = "řetězec"

spojeny_retezec = retezec_1 + " " + retezec_2 + " " + retezec_3 + " " + retezec_4
print(spojeny_retezec)

# Prázdné uvozovky s mezerami vkládáme pouze pro oddělení slov mezerou
# Kdybychom tak neudělali text by vypadal takto

spojeny_retezec = retezec_1 + retezec_2 + retezec_3 + retezec_4
print(spojeny_retezec)


Toto je spojený řetězec
Totojespojenýřetězec


## Přetypování na string
Nastanou situace kdy budeme mít datový typ např. int a chceme ho přetypovat na string. To se děje pomocí funkce str() a předáváme ji hodnotu kterou chceme přetypovat.

In [ ]:
cislo = 123
str_cislo = str(cislo)

# Nyní pokud vypíšeme datové typy obou proměnných zjistíme, že došlo k přetypování
print(cislo, str_cislo)
print(type(cislo))
print(type(str_cislo))

123 123
<class 'int'>
<class 'str'>


## Přetypování na int a float
Opačně mohou nastat situace kdy budeme mít datový typ např. string a chceme ho přetypovat na int nebo float. To se děje pomocí funkce int() a float().

In [ ]:
# mějme číselnou hodnotu zapsanou ve stringu
cislo_ve_stringu = "123"

# vypišme si datový typ této proměnné
print(type(cislo_ve_stringu))

# jak můžeme vidět jedná se o string

# nyní k samotnému přetypování
# to probíhá velmi podobně jako při převodu čísla na string
cislo_int = int(cislo_ve_stringu)
#můžeme vidět, že se došlo k přetypování
print(type(cislo_int))

# ten samý postup používáme i pro převod stringu na float
cislo_float = float(cislo_ve_stringu)
#opět můžeme vidět, že se došlo k přetypování
print(type(cislo_float))

# Kontrolní otázka: Co myslíte, že stane zde:
#text = "Ahoj"
#text_int = int(text)

## Uvozovky ve stringu
Může se stát že budeme chtít ve stringu použít uvozovky.

In [ ]:
text_bez_uvozovek = "Toto je text bez uvozovek"
text_s_uvozovkami = "Toto je text s 'uvozovkami'"
# Pokud chceme v textu uvozovky musíme zvolit jiné, než kterými určujeme datový typ string
# Pokud bychom použili stejné uvozovky, Python nám vyhodí chybu

# chybne_uvozovky = "Text s "uvozovkami"" # chyba bude vidět pokud odkomentujete tento řádek

## Předávání proměnných do stringu
Ne vždy budeme chtít vypsat přesně to co se nachází ve stringu, ale budeme text chtít pozměnit o nějakou proměnnou. Toto můžeme provést dvěma způsoby. Pomocí funkce format(), nebo pomocí f-stringu.

### Pomocí funkce format()

In [ ]:
# mějme 2 čísla
cislo_1 = 123
cislo_2 = 456

# a cheme vypsat jejich součet jako string
text = "ahoj {}"
print(text.format("Karle"))


# komu použijeme na string funkci format()
# A na místo kam chceme vložit proměnnou napíšeme {}
# Do funkce format() vložíme "cislo_1 + cislo_2", tedy operaci kterou chceme s čísli provést
# a jak si můžeme všimnout, funkce format() sama vytvoří z tohoto součtu string
print("Výsledek je: {}".format(cislo_1 + cislo_2))

# Pokud budeme chtít předat více proměnných, musíme je vložit do funkce format() v takovém pořadí, jak jsou požadovány v řetězci
string_vice_promennych = "Součet čísel {} a {} je {}".format(cislo_1, cislo_2, cislo_1 + cislo_2)
print(string_vice_promennych)

# pro lepší přehlednost můžeme do {} napsat "proměnné", na které se ve format() odkážeme

string_vice_promennych_2 = "Součet čísel {prvni} a {druhe} je {soucet}".format(prvni=cislo_1, druhe=cislo_2, soucet=cislo_1 + cislo_2)
print(string_vice_promennych_2)

### Pomocí f-stringu
Dalším způsobem předávání proměnných do stringu je pomocí f-stringu. Jediný rozdíl od funkce format(), je že proměnné zapisujeme přímo do textu. Musíme je ale opět uložit do {}.

In [ ]:
# mějme 2 čísla
cislo_1 = 123
cislo_2 = 456

a = 5
b = 7
print(a,b,a+b)
print(f"{a} {b} {a+b}")
print("{} {} {}".format(a,b,a+b))

# a cheme vypsat jejich součet jako string

# před uvozovky napíšeme "f", tím zajistíme, že se bude jednat o f-string.
# pokud budeme chtít vypsat součet čísel, napíšeme početní operaci do "{}"
# A Python se opět o převod na string z intů postará sám
print(f"Výsledek je: {cislo_1 + cislo_2}")

# Toto lze použít v jakémkoliv rozsahu
string_vice_promennych = f"Součet čísel {cislo_1} a {cislo_2} je {cislo_1 + cislo_2}"
print(string_vice_promennych)

## Manipulace se stringy
Nastanou situace kdy budeme chtít nějakým způsobem upravit string

### lower()
Pomocí funkce .lower() můžeme změnit velikost písmen v stringu na malá.

In [ ]:
text = "NěJaKÝ tExT"
print(text)

# pokud chceme změnit velikost písma, použijeme funkci lower()
mala_pismena = text.lower()
print(mala_pismena)

### split()
Pomocí funkce .split() můžeme rozdělit string na jednotlivé části. Ty se následně uloží do list.

In [ ]:
nerozdeleny_text = "Nekolik nerozdělených slov"

# jako parametr funkce split() vložíme znak, podle kterého chceme rozdělit text, defaultní znak na dělení je mezera
rozdeleny_text = nerozdeleny_text.split()
print(rozdeleny_text)

['Neko', 'ik nerozdě', 'ených s', 'ov']


### find()
Pomocí funkce .find() můžeme najít určitý znak v stringu. Pokud najdeme, funkce vrátí jeho index(číslováno od nuly). Pokud find() znak nenajde, vrátí -1.
Pokud se v textu nachází více znaků které jsme hledali, funkce find() vrátí index prvního nalezeného znaku.

In [ ]:
# máme nějaký text
text = "Nějaký text"
# a znak, který chceme najít
znak_na_najiti = "text"

#zjistíme jeho index
index_znaku = text.find(znak_na_najiti)
# vypíšeme si index
print(index_znaku)

7


### Vypisování části stringu
Python umožňuje pracovat s konkrétní částí stringu, podobně jako to je u pole.

In [ ]:
# máme nějaký text
text = "Nějaký text"

# index se kterým budeme pracovat
index = 2
# index je číslován od 0

# takto vypadá náš string rozindexovaný
# index 0 = N
# index 1 = e
# index 2 = j
# index 3 = a
# index 4 = k
# index 5 = ý
# index 6 = t
# index 7 = e
# index 8 = x
# index 9 = t


print(text)

# pokud chceme zjistit znak na konkrétní pozici, do "[]" napíšeme index, který chceme vypsat,
print(text[index])

# můžeme vypisovat část textu, která se nachází za konkrétním indexem
# znak ":" zde znamená "od", tedy vypíšeme znaky od indexu 2 do konce
print(text[index:])

# můžeme vypisovat část textu, která se nachází před konkrétním indexem
# znak ":" zde znamená "do", tedy vypíšeme znaky od začátku do indexu 2

# jediné co se změnilo bylo pořadí indexu a znak ":"
print(text[:index])

# můžeme vypisovat část textu, která se nachází mezi konkrétními indexy
# v tomto případě se vypíší znaky od indexu 2 do indexu 6
print(text[index:index + 4])

# Python umožňuje zadat do "[]" záporné číslo, tím vypíšeme konkrétní index od konce
print(text[-index])

# Kromě od-do můžeme zadat i krok
print(text[3:10:2])

# Kombinací lze např. snadno text převrátit
print(text[::-1])

Nějaký text
j
jaký text
Ně
jaký
x
aýtx
txet ýkajěN


## Cvičení
Bude následovat cvičení na procvičení tohoto notebooku. <br>
### Cvičení 1:
Dané proměnné x, y přetypujte na int, vynásobte a vložte do zadaného stringu, který předem upravíte tak, aby všechna písmena byla malá.

In [5]:
x = int(5)
y = int(50)
string_vice_promennych = f"Součet čísel x = {x} a y = {y} je {x * y}"
print(string_vice_promennych)

Součet čísel x = 5 a y = 50 je 250


Referenční řešení: <br>
<table style="margin:auto">
<tr>
<td>
<pre>
součet čísel x = 5 a y = 50 je 250
</pre>
</td>
</tr>
</table>

### Cvičení 2:
Ze zadaného textu vypište 6 znak, poslední znak (i pokud se změní délka textu) a znaky od 12 do 20 znaku.

In [3]:
string = "Text na rozkouskování"
print(string[5])
print(string[-1])
print(string[12:20])

n
í
ouskován


Referenční řešení: <br>
<table style="margin:auto">
<tr>
<td>
<pre>
n
í
ouskován
</pre>
</td>
</tr>
</table>

### Cvičení 3:
Zadanné hodnoty spojte do jednoho stringu a vypište

In [20]:
a = "Nějaký text"
s = ", který se má spojit"
d = "Přidáme k němu čísla"
f = 2
g = 6
spojeny_retezec =f" {a + s}. {d} : {f}{g}"
print(spojeny_retezec)

 Nějaký text, který se má spojit. Přidáme k němu čísla : 26


Referenční řešení: <br>
<table style="margin:auto">
<tr>
<td>
<pre>
Nějaký text, který se má spojit. Přídáme k němu čísla 26
</pre>
</td>
</tr>
</table>